pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -231,9 +236,74 @@ func (r *reconciler) getOFPriority(rule *CompletedRule) (*uint16, error) {
 	return ofPriority, nil
 }
 
+// BatchReconcile reconciles the desired state of the provided CompletedRules
+// with the actual state of Openflow entries in batch. It should only be invoked
+// if all rules are newly added without last realized status.
+func (r *reconciler) BatchReconcile(rules []*CompletedRule) error {
+	var rulesToInstall []*CompletedRule
+	var priorities []*uint16
+	for _, rule := range rules {
+		if _, exists := r.lastRealizeds.Load(rule.ID); exists {
+			klog.Errorf("rule %s already realized during the initialization phase", rule.ID)
+		} else {
+			rulesToInstall = append(rulesToInstall, rule)
+		}
+	}
+	if err := r.registerOFPriorities(rulesToInstall); err != nil {
+		return err
+	}
+	for _, rule := range rulesToInstall {
+		klog.V(2).Infof("Adding rule %s of NetworkPolicy %s/%s to be reconciled in batch", rule.ID, rule.PolicyNamespace, rule.PolicyName)
+		ofPriority, _ := r.getOFPriority(rule)
+		priorities = append(priorities, ofPriority)
+	}
+	ofRuleInstallErr := r.batchAdd(rulesToInstall, priorities)
+	if ofRuleInstallErr != nil {
+		for _, ofPriority := range priorities {
+			if ofPriority != nil {
+				r.priorityAssigner.Release(*ofPriority)
+			}
+		}
+	}
+	return ofRuleInstallErr
+}
+
+func (r *reconciler) registerOFPriorities(rules []*CompletedRule) error {
+	r.priorityMutex.Lock()
+	defer r.priorityMutex.Unlock()
+	var prioritiesToRegister []types.Priority
+	for _, r := range rules {
+		if r.PolicyPriority != nil {
+			p := types.Priority{PolicyPriority: *r.PolicyPriority, RulePriority: r.Priority}
+			prioritiesToRegister = append(prioritiesToRegister, p)
+		}
+	}
+	return r.priorityAssigner.RegisterPriorities(prioritiesToRegister)
+}
+
 // add converts CompletedRule to PolicyRule(s) and invokes installOFRule to install them.
 func (r *reconciler) add(rule *CompletedRule, ofPriority *uint16) error {
 	klog.V(2).Infof("Adding new rule %v", rule)
+	ofRuleByServicesMap, lastRealized := r.computeOFRulesForAdd(rule, ofPriority)
+	for svcHash, ofRule := range ofRuleByServicesMap {
+		// Each pod group gets an Openflow ID.
+		ofID, err := r.idAllocator.allocate()
+		if err != nil {
+			return fmt.Errorf("error allocating Openflow ID")
+		}
+		ofRule.FlowID = ofID
+		ofRule.PolicyName = lastRealized.CompletedRule.PolicyName
+		ofRule.PolicyNamespace = lastRealized.CompletedRule.PolicyNamespace
+		if err = r.installOFRule(ofRule); err != nil {
+			return err
```

Comment:
do we need to deallocate the ofID here on error?

Reply:
ofID is deallocated in installOFRule

Reply:
thanks for the clarification